gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+/**
+ * ProjectionFactor, but with pose2 (robot on the floor)
+ *
+ * This factor is useful for high-school robotics competitions,
+ * which run robots on the floor, with use fixed maps and fiducial
+ * markers.
+ *
+ * @see https://www.firstinspires.org/
+ *
+ * @file PlanarProjectionFactor.h
+ * @brief for planar smoothing
+ * @date Dec 2, 2024
+ * @author joel@truher.org
+ */
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/base/Lie.h>
+
+#include <gtsam/geometry/Cal3DS2.h>
+#include <gtsam/geometry/PinholeCamera.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose2.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/base/numericalDerivative.h>
+
+
+namespace gtsam {
+
+    /**
+     * @class PlanarProjectionFactorBase
+     * @brief Camera projection for robot on the floor.  Measurement is camera pixels.
+     */
+    class PlanarProjectionFactorBase {
+    protected:
+        PlanarProjectionFactorBase() {}
+
+        /**
+         * @param measured corresponding point in the camera frame
+         * @param poseKey index of the robot pose in the z=0 plane
+         */
+        PlanarProjectionFactorBase(const Point2& measured) : measured_(measured) {}
+
+        /**
+         * @param landmark point3
```

Comment:
Please say what the function does :-) Also, suggest to call it `predict` rather than (frowned upon) single letter.

Reply:
done